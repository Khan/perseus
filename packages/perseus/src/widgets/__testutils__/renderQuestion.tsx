import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
