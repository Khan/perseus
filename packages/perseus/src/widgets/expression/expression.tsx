import * as KAS from "@khanacademy/kas";
import {KeyArray, KeypadInput, KeypadType} from "@khanacademy/math-input";
import {linterContextDefault} from "@khanacademy/perseus-linter";
import {View} from "@khanacademy/wonder-blocks-core";
import Tooltip from "@khanacademy/wonder-blocks-tooltip";
import {LabelSmall} from "@khanacademy/wonder-blocks-typography";
import {css, StyleSheet} from "aphrodite";
import classNames from "classnames";
import * as React from "react";
import ReactDOM from "react-dom";
import _ from "underscore";

import {PerseusI18nContext} from "../../components/i18n-context";
import MathInput from "../../components/math-input";
import {useDependencies} from "../../dependencies";
import * as Changeable from "../../mixins/changeable";
import {ApiOptions, ClassNames as ApiClassNames} from "../../perseus-api";
import a11y from "../../util/a11y";

import expressionValidator from "./expression-validator";
import getDecimalSeparator from "./get-decimal-separator";

import type {DependenciesContext} from "../../dependencies";
import type {PerseusExpressionWidgetOptions} from "../../perseus-types";
import type {FocusPath, Widget, WidgetExports, WidgetProps} from "../../types";
import type {
    PerseusExpressionRubric,
    PerseusExpressionUserInput,
} from "../../validation.types";
import type {Keys as Key, KeypadConfiguration} from "@khanacademy/math-input";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type InputPath = ReadonlyArray<string>;

// Map of international operator names to their English equivalents
const englishOperators = {
    arctg: "arctan",
    cosec: "csc",
    cossec: "csc",
    cotg: "cot",
    ctg: "cot",
    sen: "sin",
    tg: "tan",
};

const anglicizeOperators = (tex: string): string => {
    // sen is used instead of sin in some languages, e.g. Portuguese.
    // To ensure that answers in various languages are graded correctly, we
    // convert operators to their Englishy forms.
    return tex.replace(
        /\\operatorname{([a-z]+)}/g,
        (_, op) => `\\${englishOperators[op] ?? op} `,
    );
};

const normalizeTex = (tex: string): string => {
    return anglicizeOperators(tex);
};

type RenderProps = {
    buttonSets: PerseusExpressionWidgetOptions["buttonSets"];
    buttonsVisible?: PerseusExpressionWidgetOptions["buttonsVisible"];
    functions: PerseusExpressionWidgetOptions["functions"];
    times: PerseusExpressionWidgetOptions["times"];
    visibleLabel: PerseusExpressionWidgetOptions["visibleLabel"];
    ariaLabel: PerseusExpressionWidgetOptions["ariaLabel"];
    keypadConfiguration: ReturnType<typeof keypadConfigurationForProps>;
};

type ExternalProps = WidgetProps<RenderProps, PerseusExpressionRubric>;

export type Props = ExternalProps &
    Partial<React.ContextType<typeof DependenciesContext>> & {
        apiOptions: NonNullable<ExternalProps["apiOptions"]>;
        buttonSets: NonNullable<ExternalProps["buttonSets"]>;
        functions: NonNullable<ExternalProps["functions"]>;
        linterContext: NonNullable<ExternalProps["linterContext"]>;
        onBlur: NonNullable<ExternalProps["onBlur"]>;
        onFocus: NonNullable<ExternalProps["onFocus"]>;
        times: NonNullable<ExternalProps["times"]>;
        visibleLabel: PerseusExpressionWidgetOptions["visibleLabel"];
        ariaLabel: PerseusExpressionWidgetOptions["ariaLabel"];
        value: string;
    };

export type ExpressionState = {
    invalid: boolean;
    showErrorTooltip: boolean;
    showErrorStyle: boolean;
};

type DefaultProps = {
    apiOptions: Props["apiOptions"];
    buttonSets: Props["buttonSets"];
    functions: Props["functions"];
    linterContext: Props["linterContext"];
    onBlur: Props["onBlur"];
    onFocus: Props["onFocus"];
    times: Props["times"];
    value: Props["value"];
};

// The new, MathQuill input expression widget
export class Expression
    extends React.Component<Props, ExpressionState>
    implements Widget
{
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    _textareaId = `expression_textarea_${Date.now()}`;
    _isMounted = false;

    static getUserInputFromProps(props: Props): PerseusExpressionUserInput {
        return normalizeTex(props.value);
    }

    static defaultProps: DefaultProps = {
        value: "",
        times: false,
        functions: [],
        buttonSets: ["basic", "trig", "prealgebra", "logarithms"],
        onFocus: () => {},
        onBlur: () => {},
        apiOptions: ApiOptions.defaults,
        linterContext: linterContextDefault,
    };

    displayName = "Expression";

    state: ExpressionState = {
        invalid: false,
        showErrorTooltip: false,
        showErrorStyle: false,
    };

    componentDidMount: () => void = () => {
        document.addEventListener("mousedown", this._handleMouseDown);

        // TODO(scottgrant): This is a hack to remove the deprecated call to
        // this.isMounted() but is still considered an anti-pattern.
        this._isMounted = true;

        // HACK: imperatively add an ID onto the Mathquill input
        // (which in mobile is a span; desktop a textarea)
        // in order to associate a visual label with it
        if (this.refs.input) {
            const isMobile = this.props.apiOptions.customKeypad;
            const container = ReactDOM.findDOMNode(this.refs.input);
            const selector = isMobile ? ".mq-textarea > span" : "textarea";
            const inputElement = (container as Element).querySelector(selector);
            inputElement?.setAttribute("id", this._textareaId);
        }
    };

    // Whenever the input value changes, attempt to parse it.
    //
    // Clear any errors if this parse succeeds, show an error within a second
    // if it fails.
    componentDidUpdate: (prevProps: Props) => void = (prevProps) => {
        if (
            !_.isEqual(this.props.value, prevProps.value) ||
            !_.isEqual(this.props.functions, prevProps.functions)
        ) {
            this.setState({
                invalid: false,
                showErrorTooltip: false,
                showErrorStyle: false,
            });
            if (!this.parse(this.props.value, this.props).parsed) {
                this.setState({
                    invalid: true,
                });
            }
        }
    };

    componentWillUnmount: () => void = () => {
        this._isMounted = false;
    };

    _handleMouseDown = () => {
        if (this._isMounted && this.state.showErrorTooltip) {
            this.setState({
                showErrorTooltip: false,
            });
        }
    };

    getUserInput(): PerseusExpressionUserInput {
        return Expression.getUserInputFromProps(this.props);
    }

    change: (...args: any) => any | undefined = (...args: any) => {
        return Changeable.change.apply(this, args);
    };

    parse: (value: string, props: Props) => any = (
        value: string,
        props: Props,
    ) => {
        // TODO(jack): Disable icu for content creators here, or
        // make it so that solution answers with ','s or '.'s work
        const options = _.pick(props || this.props, "functions");
        _.extend(options, {
            decimal_separator: getDecimalSeparator(this.context.locale),
        });
        return KAS.parse(normalizeTex(value), options);
    };

    changeAndTrack: (e: any, cb: () => void) => void = (
        e: any,
        cb: () => void,
    ) => {
        this.change("value", e, cb);
        this.props.trackInteraction();
    };

    _handleFocus: () => void = () => {
        this.props.analytics?.onAnalyticsEvent({
            type: "perseus:expression-focused",
            payload: null,
        });

        /* c8 ignore next */
        this.props.onFocus([]);
    };

    _handleBlur: () => void = () => {
        /* c8 ignore next */
        this.props.onBlur([]);
    };

    focus: () => boolean = () => {
        if (this.props.apiOptions.customKeypad) {
            // eslint-disable-next-line react/no-string-refs
            // @ts-expect-error - TS2339 - Property 'focus' does not exist on type 'ReactInstance'.
            this.refs.input.focus();
        } else {
            // The buttons are often on top of text you're trying to read, so
            // don't focus the editor automatically.
        }

        return true;
    };

    focusInputPath(inputPath: InputPath) {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'focus' does not exist on type 'ReactInstance'.
        this.refs.input.focus();
    }

    blurInputPath(inputPath: InputPath) {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'blur' does not exist on type 'ReactInstance'.
        if (typeof this.refs.input?.blur === "function") {
            // eslint-disable-next-line react/no-string-refs
            // @ts-expect-error - TS2339 - Property 'blur' does not exist on type 'ReactInstance'.
            this.refs.input?.blur();
        }
    }

    // HACK(joel)
    insert(keyPressed: Key) {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'insert' does not exist on type 'ReactInstance'.
        this.refs.input.insert(keyPressed);
    }

    getInputPaths: () => ReadonlyArray<ReadonlyArray<any>> = () => {
        // The widget itself is an input, so we return a single empty list to
        // indicate this.
        /* c8 ignore next */
        return [[]];
    };

    // TODO(Nicole): I believe this is going away and won't be needed anymore
    getGrammarTypeForPath(inputPath: FocusPath): string {
        /* c8 ignore next */
        return "expression";
    }

    setInputValue(path: FocusPath, newValue: string, cb: () => void) {
        this.props.onChange(
            {
                value: newValue,
            },
            cb,
        );
    }

    render() {
        if (this.props.apiOptions.customKeypad) {
            return (
                <View className={css(styles.mobileLabelInputWrapper)}>
                    {!!this.props.visibleLabel && (
                        <LabelSmall htmlFor={this._textareaId} tag="label">
                            {this.props.visibleLabel}
                        </LabelSmall>
                    )}
                    <KeypadInput
                        // eslint-disable-next-line react/no-string-refs
                        ref="input"
                        ariaLabel={
                            this.props.ariaLabel ||
                            this.context.strings.mathInputBox
                        }
                        value={this.props.value}
                        keypadElement={this.props.keypadElement}
                        onChange={this.changeAndTrack}
                        onFocus={() => {
                            // this.props.keypadElement should always be set
                            // when apiOptions.customKeypad is set, but how
                            // to convince TypeScript of this?
                            this.props.keypadElement?.configure(
                                this.props.keypadConfiguration,
                                () => {
                                    if (this._isMounted) {
                                        this._handleFocus();
                                    }
                                },
                            );
                        }}
                        onBlur={this._handleBlur}
                    />
                </View>
            );
        }

        const className = classNames({
            "perseus-widget-expression": true,
            "show-error-tooltip": this.state.showErrorTooltip,
        });

        const {ERROR_MESSAGE, ERROR_TITLE} = this.context.strings;

        return (
            <View className={css(styles.desktopLabelInputWrapper)}>
                {!!this.props.visibleLabel && (
                    <LabelSmall htmlFor={this._textareaId} tag="label">
                        {this.props.visibleLabel}
                    </LabelSmall>
                )}
                <div
                    className={className}
                    onBlur={() =>
                        this.state.invalid &&
                        this.setState({
                            showErrorTooltip: true,
                            showErrorStyle: true,
                        })
                    }
                    onFocus={() =>
                        this.setState({
                            showErrorTooltip: false,
                        })
                    }
                >
                    {/**
                * This is a visually hidden container for the error tooltip.
                https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/alert_role#example_3_visually_hidden_alert_container_for_screen_reader_notifications
            */}
                    <View style={a11y.srOnly} role="alert">
                        {this.state.showErrorTooltip &&
                            ERROR_TITLE + " " + ERROR_MESSAGE}
                    </View>
                    <Tooltip
                        forceAnchorFocusivity={false}
                        opened={this.state.showErrorTooltip}
                        title={ERROR_TITLE}
                        content={ERROR_MESSAGE}
                    >
                        <MathInput
                            // eslint-disable-next-line react/no-string-refs
                            ref="input"
                            className={ApiClassNames.INTERACTIVE}
                            value={this.props.value}
                            onChange={this.changeAndTrack}
                            convertDotToTimes={this.props.times}
                            buttonSets={this.props.buttonSets}
                            onFocus={this._handleFocus}
                            onBlur={this._handleBlur}
                            hasError={this.state.showErrorStyle}
                            ariaLabel={
                                this.props.ariaLabel ||
                                this.context.strings.mathInputBox
                            }
                            extraKeys={
                                this.props.keypadConfiguration?.extraKeys
                            }
                            analytics={
                                this.props.analytics ?? {
                                    onAnalyticsEvent: async () => {},
                                }
                            }
                        />
                    </Tooltip>
                </div>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    mobileLabelInputWrapper: {
        padding: "15px 4px 0",
    },
    desktopLabelInputWrapper: {
        margin: "5px 5px 0",
    },
});

/**
 * Determine the keypad configuration parameters for the input, based on the
 * provided properties.
 *
 * There are two configuration parameters to be passed to the keypad:
 *   (1) The keypad type. For the Expression widget, we always use the
 *       Expression keypad.
 *   (2) The extra keys; namely, any variables or constants (like Pi) that need
 *       to be included as keys on the keypad. These are scraped from the answer
 *       forms.
 */
export const keypadConfigurationForProps = (
    widgetOptions: PerseusExpressionWidgetOptions,
): KeypadConfiguration => {
    // Always use the Expression keypad, regardless of the button sets that have
    // been enabled.
    const keypadType = KeypadType.EXPRESSION;

    // Extract any and all variables and constants from the answer forms.
    const uniqueExtraVariables: Partial<Record<Key, boolean>> = {};
    const uniqueExtraConstants: Partial<Record<Key, boolean>> = {};
    for (const answerForm of widgetOptions.answerForms) {
        const maybeExpr = KAS.parse(answerForm.value, widgetOptions);
        if (maybeExpr.parsed) {
            const expr = maybeExpr.expr;

            // The keypad expects Greek letters to be capitalized (e.g., it
            // requires `PI` instead of `pi`). Right now, it only supports Pi
            // and Theta, so we special-case.
            const isGreek = (symbol: any) =>
                symbol === "pi" || symbol === "theta";
            const toKey = (symbol: any) =>
                isGreek(symbol) ? symbol.toUpperCase() : symbol;
            const isKey = (key: string): key is Key =>
                KeyArray.includes(key as Key);

            for (const variable of expr.getVars()) {
                const maybeKey = toKey(variable);
                if (isKey(maybeKey)) {
                    uniqueExtraVariables[maybeKey] = true;
                }
            }
            for (const constant of expr.getConsts()) {
                const maybeKey = toKey(constant);
                if (isKey(maybeKey)) {
                    uniqueExtraConstants[maybeKey] = true;
                }
            }
        }
    }

    // TODO(charlie): Alert the keypad as to which of these symbols should be
    // treated as functions.
    const extraVariables = Object.keys(
        uniqueExtraVariables,
    ).sort() as ReadonlyArray<Key>;

    const extraConstants = Object.keys(
        uniqueExtraConstants,
    ).sort() as ReadonlyArray<Key>;

    let extraKeys = [...extraVariables, ...extraConstants];
    if (!extraKeys.length) {
        // If there are no extra symbols available, we include Pi anyway, so
        // that the "extra symbols" button doesn't appear empty.
        extraKeys = ["PI"];
    }

    return {
        keypadType,
        extraKeys,
        times: widgetOptions.times,
    };
};

const propUpgrades = {
    /* c8 ignore next */
    "1": (v0props: any): PerseusExpressionWidgetOptions => ({
        times: v0props.times,
        buttonSets: v0props.buttonSets,
        functions: v0props.functions,
        buttonsVisible: v0props.buttonsVisible,
        visibleLabel: v0props.visibleLabel,
        ariaLabel: v0props.ariaLabel,

        answerForms: [
            {
                considered: "correct",
                form: v0props.form,
                simplify: v0props.simplify,
                value: v0props.value,
            },
        ],
    }),
} as const;

const ExpressionWithDependencies = React.forwardRef<
    Expression,
    Omit<PropsFor<typeof Expression>, keyof ReturnType<typeof useDependencies>>
>((props, ref) => {
    const deps = useDependencies();
    return <Expression ref={ref} analytics={deps.analytics} {...props} />;
});

// HACK: Propogate "static" methods onto our wrapper component.
// In the future we should adjust client apps to not depend on these static
// methods and instead adjust Perseus to provide these facilities through
// instance methods on our Renderers.
// @ts-expect-error - TS2339 - Property 'validate' does not exist on type
ExpressionWithDependencies.getUserInputFromProps =
    Expression.getUserInputFromProps;

export default {
    name: "expression",
    displayName: "Expression / Equation",
    accessible: true,
    defaultAlignment: "inline-block",
    widget: ExpressionWithDependencies,
    transform: (widgetOptions: PerseusExpressionWidgetOptions): RenderProps => {
        const {
            times,
            functions,
            buttonSets,
            buttonsVisible,
            visibleLabel,
            ariaLabel,
        } = widgetOptions;
        return {
            keypadConfiguration: keypadConfigurationForProps(widgetOptions),
            times,
            functions,
            buttonSets,
            buttonsVisible,
            visibleLabel,
            ariaLabel,
        };
    },
    version: {major: 1, minor: 0},
    propUpgrades: propUpgrades,

    // For use by the editor
    isLintable: true,
    validator: expressionValidator,

    getOneCorrectAnswerFromRubric(
        rubric: PerseusExpressionRubric,
    ): string | null | undefined {
        const correctAnswers = (rubric.answerForms || []).filter(
            (answerForm) => answerForm.considered === "correct",
        );
        if (correctAnswers.length === 0) {
            return;
        }
        return correctAnswers[0].value;
    },
} as WidgetExports<typeof ExpressionWithDependencies>;
