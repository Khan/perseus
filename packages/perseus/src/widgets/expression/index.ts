export {default, Expression} from "./expression";
