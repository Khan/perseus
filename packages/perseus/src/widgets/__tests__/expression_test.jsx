// @flow
import {screen, fireEvent} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import {testDependencies} from "../../../../../testing/test-dependencies.js";
import * as Dependencies from "../../dependencies.js";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing.js";
import {
    expressionItem2,
    expressionItem3,
    expressionItem3Options,
} from "../__testdata__/expression_testdata.js";
import {Expression} from "../expression.jsx";

import {renderQuestion} from "./renderQuestion.jsx";

const assertComplete = (itemData, input, isCorrect) => {
    const {renderer} = renderQuestion(itemData.question);
    userEvent.type(screen.getByRole("textbox"), input);
    const [_, score] = renderer.guessAndScore();
    expect(score).toMatchObject({
        type: "points",
        earned: isCorrect ? 1 : 0,
    });
};

const assertCorrect = (itemData, input) =>
    assertComplete(itemData, input, true);
const assertIncorrect = (itemData, input) =>
    assertComplete(itemData, input, false);

// TODO: actually Assert that message is being set on the score object.
const assertInvalid = (itemData, input, message?: string) => {
    const {renderer} = renderQuestion(itemData.question);
    if (input.length) {
        userEvent.type(screen.getByRole("textbox"), input);
    }
    const [_, score] = renderer.guessAndScore();
    expect(score).toMatchObject({type: "invalid"});
};

describe("Expression Widget", function () {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
        registerAllWidgetsForTesting();
    });

    describe("grading", function () {
        it("should not grade a thing that doesn't parse", function () {
            assertInvalid(expressionItem2, "+++");
        });

        it("should not grade a thing that is empty", function () {
            assertInvalid(expressionItem2, "");
        });
    });

    describe("fallthrough", function () {
        it("should grade answers which don't match anything as wrong", function () {
            assertIncorrect(expressionItem2, "500");
        });
    });

    describe("variables", function () {
        it.each(["123-X", "X-123"])(
            "should not grade answers that are correct except for the " +
                "variable case",
            function (input) {
                assertInvalid(expressionItem2, input);
            },
        );

        // This isn't necessarily the ideal behavior; it's just what was the
        // easiest to implement
        it(
            "should not grade answers that have the wrong variable case, " +
                "even if the answer has got other errors",
            function () {
                assertInvalid(expressionItem2, "123+X");
            },
        );

        it.each(["123-y", "123-Y"])(
            "should not not grade answers that use the wrong variable",
            function (input) {
                assertInvalid(expressionItem2, input);
            },
        );
    });

    describe("multiple answers", function () {
        it.each(["x-123", "123-x"])(
            "should recognize either of two possibilities",
            (input) => {
                assertCorrect(expressionItem2, input);
            },
        );

        /* For the tests below, here are the three answer forms:
         *
         *     considered: "ungraded"
         *     value: x+1
         *
         *     considered: "incorrect"
         *     value: y+1
         *
         *     considered: "correct"
         *     value: z+1
         */

        it("should match from top to bottom", function () {
            assertInvalid(expressionItem3, "x+1");
        });

        it("should match from top to bottom (2)", function () {
            assertIncorrect(expressionItem3, "y+1");
        });

        it("should match from top to bottom (3)", function () {
            assertCorrect(expressionItem3, "z+1");
        });

        it.each([["X+1"], ["Y+1"], ["Z+1"]])(
            "should give casing or variable name error only relative to the " +
                "correct answer",
            (input) => {
                assertInvalid(expressionItem3, input);
            },
        );
    });

    describe("ungraded", function () {
        it("should handle ungraded answers with no error callback", function () {
            const err = Expression.validate("x+1", expressionItem3Options);
            expect(err).toStrictEqual({message: "", type: "invalid"});
        });
    });

    describe("invalid", function () {
        it("should handle ungraded answers with no error callback", function () {
            const err = Expression.validate("x+^1", expressionItem3Options);
            expect(err).toStrictEqual({message: null, type: "invalid"});
        });
    });
});

describe("getOneCorrectAnswerFromRubric", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should return undefined if rubric.value is null/undefined", () => {
        // Arrange
        const rubric = {
            answerForms: [],
            buttonSets: [],
            functions: [],
            times: true,
        };

        // Act
        const result = Expression.getOneCorrectAnswerFromRubric(rubric);

        // Assert
        expect(result).toBeUndefined();
    });

    it("returns a correct answer when there is one correct answer", () => {
        // Arrange
        const rubric = {
            answerForms: [
                {
                    value: "123",
                    form: true,
                    simplify: false,
                    considered: "correct",
                },
            ],
            buttonSets: [],
            functions: [],
            times: true,
        };

        // Act
        const result = Expression.getOneCorrectAnswerFromRubric(rubric);

        // Assert
        expect(result).toEqual("123");
    });

    it("returns the first correct answer when there are multiple correct answers", () => {
        // Arrange
        const rubric = {
            answerForms: [
                {
                    value: "123",
                    form: true,
                    simplify: false,
                    considered: "correct",
                },
                {
                    value: "456",
                    form: true,
                    simplify: false,
                    considered: "correct",
                },
            ],
            buttonSets: [],
            functions: [],
            times: true,
        };

        // Act
        const result = Expression.getOneCorrectAnswerFromRubric(rubric);

        // Assert
        expect(result).toEqual("123");
    });
});

describe("focus state", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("supports directly focusing", () => {
        //  Arrange
        renderQuestion(expressionItem2.question);

        // Act
        const expressionInput = screen.getByRole("textbox");
        expressionInput.focus();

        // Assert
        expect(expressionInput).toHaveFocus();
    });

    it("supports directly blurring", () => {
        //  Arrange
        renderQuestion(expressionItem2.question);

        // Act
        const expressionInput = screen.getByRole("textbox");
        expressionInput.focus();
        expressionInput.blur();

        // Assert
        expect(expressionInput).not.toHaveFocus();
    });

    it("can be focused via a function", () => {
        // arrange
        const {renderer} = renderQuestion(expressionItem2.question);
        const expression = renderer.findWidgets("expression 1")[0];

        // act
        expression.focusInputPath();

        // Assert
        const expressionInput = screen.getByRole("textbox");
        expect(expressionInput).toHaveFocus();
    });
});

describe("rendering", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("supports mobile rendering", async () => {
        // arrange and act
        renderQuestion(expressionItem2.question, {
            // Setting this triggers mobile rendering
            // it would be nice if this was more clear in the code
            customKeypad: true,
        });

        // Assert
        const mobileInput = await screen.findByRole("textbox");
        expect(mobileInput).toBeVisible();
    });

    it("static renders", () => {
        // arrange and act
        const {container} = renderQuestion(expressionItem2.question, {
            staticRender: true,
        });

        // Assert
        expect(container).toMatchSnapshot("static render");
    });
});

describe("interaction", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("sets input value directly", () => {
        // arrange
        const {renderer} = renderQuestion(expressionItem2.question);

        // act
        renderer.setInputValue(["expression 1"], "123-x", () => {});
        const score = renderer.guessAndScore()[1];

        // Assert
        expect(score.type).toBe("points");
        // Score.total doesn't exist if the input is invalid
        // In this case we know that it'll be valid so we can assert directly
        // $FlowFixMe[prop-missing]
        expect(score.earned).toBe(score.total);
    });

    it("has a developer facility for inserting", () => {
        // arrange
        const {renderer} = renderQuestion(expressionItem2.question);
        const expression = renderer.findWidgets("expression 1")[0];
        expression.insert("x+1");

        // act
        const score = renderer.score();

        // Assert
        // Score.total doesn't exist if the input is invalid
        // In this case we know that it'll be valid so we can assert directly
        // $FlowFixMe[prop-missing]
        expect(score.total).toBe(1);
    });
});

describe("error tooltip", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("shows on error", () => {
        // Arrange
        const {renderer} = renderQuestion(expressionItem2.question);
        const expression = renderer.findWidgets("expression 1")[0];

        // Act
        expression.insert("x&&&&&^1");
        renderer.guessAndScore();
        jest.runOnlyPendingTimers();

        // Assert
        const errorMessage = screen.getByText(
            "Sorry, I don't understand that!",
        );
        expect(errorMessage).not.toBeUndefined();
    });

    it("shows error text on mouse over", async () => {
        // Arrange
        const {renderer} = renderQuestion(expressionItem2.question);
        const expression = renderer.findWidgets("expression 1")[0];

        // Act
        expression.insert("x&&&&&^1");
        renderer.guessAndScore();
        jest.runOnlyPendingTimers();

        const tooltip = await screen.findByTestId("test-error-icon");

        // eslint-disable-next-line testing-library/prefer-user-event
        fireEvent.mouseEnter(tooltip);

        // Assert
        expect(
            screen.getByText("Sorry, I don't understand that!"),
        ).toBeVisible();
    });
    it("hides error text on mouse leave", async () => {
        // Arrange
        const {renderer} = renderQuestion(expressionItem2.question);
        const expression = renderer.findWidgets("expression 1")[0];

        expression.insert("x&&&&&^1");
        renderer.guessAndScore();
        jest.runOnlyPendingTimers();

        const tooltip = await screen.findByTestId("test-error-icon");

        // Act
        // NOTE(Nicole): The existing perseus code requires explicit events
        // eslint-disable-next-line testing-library/prefer-user-event
        fireEvent.mouseEnter(tooltip);
        // NOTE(Nicole): The existing perseus code requires explicit events
        // eslint-disable-next-line testing-library/prefer-user-event
        fireEvent.mouseLeave(tooltip);

        // Assert
        expect(
            screen.getByText("Sorry, I don't understand that!"),
        ).not.toBeVisible();
    });

    it.skip("toggles error text on click", async () => {
        // Arrange
        const {renderer} = renderQuestion(expressionItem2.question);
        const expression = renderer.findWidgets("expression 1")[0];

        expression.insert("x&&&&&^1");
        renderer.guessAndScore();
        jest.runOnlyPendingTimers();

        const tooltip = await screen.findByTestId("test-error-icon");

        // Act & Assert
        // NOTE(Nicole): The existing perseus code requires explicit events
        // eslint-disable-next-line testing-library/prefer-user-event
        fireEvent.click(tooltip);
        expect(
            screen.getByText("Sorry, I don't understand that!"),
        ).toBeVisible();

        // Act & Assert
        // NOTE(Nicole): The existing perseus code requires explicit events
        // eslint-disable-next-line testing-library/prefer-user-event
        fireEvent.click(tooltip);
        expect(
            screen.getByText("Sorry, I don't understand that!"),
        ).not.toBeVisible();
    });
});
