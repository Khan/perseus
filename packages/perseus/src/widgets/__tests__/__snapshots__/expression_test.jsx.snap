// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering static renders: static render 1`] = `
<div>
  <span
    class="mock-KatexProvider"
  >
    <div
      class="perseus-renderer perseus-renderer-responsive"
    >
      <div
        class="paragraph"
        data-perseus-paragraph-index="0"
      >
        <div
          class="paragraph"
        >
          <div
            class="perseus-widget-container widget-nohighlight widget-inline-block"
          >
            <span
              class="math-output perseus-input perseus-interactive"
            >
              <div
                style="text-align: center;"
              >
                <span
                  class="mock-TeX"
                />
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;
